src/sync/stream.ts

```diff
@@ -0,0 +1,486 @@
+   */
+  updateUI() {
+    // Queue the UI update to run after we process the event that
+    // generated this update.
+    this.executionQueue = this.executionQueue.then(async () => {
+      if (this.signal.aborted) {
+        return;
+      }
+      if (this.canUpdateUI) {
+        this.canUpdateUI = false;
+        this.connection.emit('wallet-load-partial-update', {
+          addressesFound: await this.storage.store.addressCount(),
+          historyLength: await this.storage.store.historyCount(),
+        });
+        setTimeout(() => {
+          this.canUpdateUI = true;
+        }, this.UI_UPDATE_INTERVAL);
+      }
+    });
+  }
+
+  sendStartMessage() {
+    switch (this.mode) {
+      case HistorySyncMode.XPUB_STREAM_WS:
+        this.connection.sendStartXPubStreamingHistory(
+          this.streamId,
+          this.lastLoadedIndex + 1,
+          this.xpubkey,
+          this.gapLimit
+        );
+        break;
+      case HistorySyncMode.MANUAL_STREAM_WS:
+        this.connection.sendManualStreamingHistory(
+          this.streamId,
+          this.lastLoadedIndex + 1,
+          loadAddressesCPUIntensive(
+            this.lastLoadedIndex + 1,
+            this.ADDRESSES_PER_MESSAGE,
+            this.xpubkey,
+            this.network
+          ),
+          true,
+          this.gapLimit
+        );
+        this.lastLoadedIndex += this.ADDRESSES_PER_MESSAGE;
+        break;
+      default:
+        // Should never happen.
+        this.abortWithError('Unknown stream mode');
+        break;
+    }
+  }
+
+  async shutdown() {
+    await this.executionQueue;
+    await this.batchQueue;
+
+    if (this.errorMessage) {
+      throw new Error(this.errorMessage);
+    }
+  }
+}
+
+function buildListener(manager: StreamManager, resolve: () => void) {
+  return (wsData: IStreamSyncHistoryData) => {
+    // Early return if the stream is aborted
+    // This will prevent any tx or address from being added when we want to stop the stream.
+    if (manager.signal.aborted) {
+      return;
+    }
+    // Only process the message if it is from our stream, this error should not happen.
+    if (wsData.id !== manager.streamId) {
+      // Check that the stream id is the same we sent
+      return;
+    }
+    // Vertex is a transaction in the history of the last address received
+    if (isStreamSyncHistoryVertex(wsData)) {
+      // foundAnyTx = true;
+      // add to history
+      manager.addTx(wsData.data);
+      manager.updateUI();
+    }
+    if (isStreamSyncHistoryAddress(wsData)) {
+      manager.addAddress(wsData.index, wsData.address);
+      manager.updateUI();
+      manager.generateNextBatch();
+    }
+    if (isStreamSyncHistoryEnd(wsData)) {
+      // cleanup and stop the method.
+      resolve();
+    }
+    // An error happened on the fullnode, we should stop the stream
+    if (isStreamSyncHistoryError(wsData)) {
+      console.error(`Stream error: ${wsData.errmsg}`);
+      manager.abortWithError(wsData.errmsg);
+    }
```

Comment:
Isn't it better to use the following structure:

```
if (a) {
} else if (b) {
} else if (c) {
} else {
    throw new "should never happen";
}

Reply:
Done

Reply:
Thanks!